File: genshin-builder-mobile/lib/data/amber/amber_upgrade.dart
import 'package:http/http.dart' as http;

import '../../domain/models/calculation_models.dart';
import '../../domain/weapon_exp.dart';
import '../amber/amber_constants.dart';
import '../config/remote_json_fetch.dart';

const _expMaterialIds = <String, List<String>>{
  'character': ['104001', '104002', '104003'],
  'weapon': ['104011', '104012', '104013'],
};

const _concurrency = 4;
const _requestDelayMs = 100;

/// Project Amber v2 から突破・天賦・武器 upgrade を取得（Web `amber-upgrade.ts` 相当）
class AmberUpgradeApi {
  AmberUpgradeApi({http.Client? client}) : _client = client ?? http.Client();

  final http.Client _client;
  static const _timeout = Duration(seconds: 20);
  static const _maxBodyBytes = 2 * 1024 * 1024;

  Future<Map<String, dynamic>> _fetchV2(String path) async {
    final json = await fetchRemoteJsonMap(
      client: _client,
      url: '$amberBaseUrl/api/v2/jp$path',
      kind: 'amber-upgrade',
      timeout: _timeout,
      maxBytes: _maxBodyBytes,
    );
    if (json['response'] != 200) {
      throw const FormatException('Invalid Amber response status');
    }
    final data = json['data'];
    if (data is! Map<String, dynamic>) {
      throw const FormatException('Invalid Amber upgrade payload');
    }
    return data;
  }

  List<PromoteStage> parsePromotes(List<dynamic>? promotes) {
    return (promotes ?? []).map((raw) {
      if (raw is! Map<String, dynamic>) {
        throw const FormatException('Invalid promote entry');
      }
      final p = raw;
      final costItems = <String, int>{};
      final costs = p['costItems'];
      if (costs is Map) {
        costs.forEach((key, value) {
          final count = (value as num?)?.toInt();
          if (count == null || count < 0) {
            throw const FormatException('Invalid promote material count');
          }
          costItems['$key'] = count;
        });
      } else if (costs is List) {
        for (final item in costs) {
          if (item is! Map<String, dynamic>) {
            throw const FormatException('Invalid promote material');
          }
          final id = '${item['id']}';
          final count = (item['count'] as num?)?.toInt();
          if (id.isEmpty || id == 'null' || count == null || count < 0) {
            throw const FormatException('Invalid promote material');
          }
          costItems[id] = count;
        }
      }
      final promoteLevel = (p['promoteLevel'] as num?)?.toInt() ?? 0;
      final unlockMaxLevel = (p['unlockMaxLevel'] as num?)?.toInt() ?? 90;
      final coinCost = (p['coinCost'] as num?)?.toInt() ?? 0;
      if (promoteLevel < 0 || unlockMaxLevel < 1 || coinCost < 0) {
        throw const FormatException('Invalid promote values');
      }
      return PromoteStage(
        promoteLevel: promoteLevel,
        unlockMaxLevel: unlockMaxLevel,
        costItems: costItems,
        coinCost: coinCost,
        requiredPlayerLevel: (p['requiredPlayerLevel'] as num?)?.toInt(),
      );
    }).toList();
  }

  List<TalentLevelUpgrade> parseTalentUpgrades(Map<String, dynamic>? promote) {
    if (promote == null) return [];
    final entries =
        promote.entries
            .map((e) => e.value as Map<String, dynamic>)
            .where((p) => p['level'] != null)
            .toList()
          ..sort(
            (a, b) =>
                ((a['level'] as num?)?.toInt() ?? 0) -
                ((b['level'] as num?)?.toInt() ?? 0),
          );

    return entries.map((p) {
      final costItems = <String, int>{};
      final costs = p['costItems'];
      if (costs is Map) {
        costs.forEach((key, value) {
          final count = (value as num?)?.toInt();
          if (count == null || count < 0) {
            throw const FormatException('Invalid talent material count');
          }
          costItems['$key'] = count;
        });
      }
      final level = (p['level'] as num?)?.toInt() ?? 0;
      final coinCost = (p['coinCost'] as num?)?.toInt() ?? 0;
      if (level < 1 || coinCost < 0) {
        throw const FormatException('Invalid talent values');
      }
      return TalentLevelUpgrade(
        level: level,
        costItems: costItems,
        coinCost: coinCost,
      );
    }).toList();
  }

  Future<
    ({
      List<PromoteStage> promotes,
      Map<String, List<TalentLevelUpgrade>> talents,
    })?
  >
  fetchCharacterUpgrade(String characterId) async {
    final data = await _fetchV2('/avatar/$characterId');

    if (data['talent'] is! Map<String, dynamic> ||
        data['upgrade'] is! Map<String, dynamic>) {
      throw const FormatException('Missing character upgrade fields');
    }
    final talentRaw = data['talent'] as Map<String, dynamic>;
    final keys =
        talentRaw.keys.toList()
          ..sort((a, b) => int.parse(a).compareTo(int.parse(b)));

    final talents = <String, List<TalentLevelUpgrade>>{};
    var skillIndex = 0;
    for (final key in keys) {
      final t = talentRaw[key] as Map<String, dynamic>;
      final type = (t['type'] as num?)?.toInt();
      if (type != 0 && type != 1) continue;
      final upgrades = parseTalentUpgrades(
        t['promote'] as Map<String, dynamic>?,
      );
      if (upgrades.isNotEmpty) {
        talents['skill_$skillIndex'] = upgrades;
        skillIndex++;
      }
      if (skillIndex >= 3) break;
    }

    final upgrade = data['upgrade'] as Map<String, dynamic>;
    final promotes = parsePromotes(upgrade['promote'] as List<dynamic>?);
    if (promotes.isEmpty) {
      throw const FormatException('Empty character promotes');
    }

    return (promotes: promotes, talents: talents);
  }

  Future<({List<PromoteStage> promotes, List<String> levelUpItemIds})?>
  fetchWeaponUpgrade(String weaponId) async {
    final data = await _fetchV2('/weapon/$weaponId');

    // マネキン武器など upgrade 未実装はスキップ（同期全体を落とさない）
    if (data['upgrade'] is! Map<String, dynamic>) {
      return null;
    }
    final upgrade = data['upgrade'] as Map<String, dynamic>;
    final promoteRaw = upgrade['promote'];
    if (promoteRaw is! List) {
      return null;
    }
    final promotes = parsePromotes(promoteRaw);
    if (promotes.isEmpty) {
      return null;
    }
    final items = data['items'] as Map<String, dynamic>?;
    final levelUpItemIds = parseWeaponEnhancementOreIds(items);

    return (promotes: promotes, levelUpItemIds: levelUpItemIds);
  }

  int? parseExpFromDescription(String description) {
    final match = RegExp(r'経験値(\d+)').firstMatch(description);
    return match == null ? null : int.tryParse(match.group(1)!);
  }

  Future<List<({String materialId, int exp, String targetType})>>
  fetchLevelUpMaterials() async {
    final result = <({String materialId, int exp, String targetType})>[];

    for (final targetType in ['character', 'weapon']) {
      for (final id in _expMaterialIds[targetType]!) {
        final detail = await _fetchV2('/material/$id');
        final description = detail['description'] as String? ?? '';
        final exp = parseExpFromDescription(description);
        if (exp == null) continue;
        result.add((materialId: id, exp: exp, targetType: targetType));
      }
    }

    return result;
  }

  Future<List<T>> mapWithConcurrency<T>(
    List<String> ids,
    Future<T?> Function(String id) fn, {
    void Function(int completed, int total)? onProgress,
  }) async {
    final results = <T>[];
    var index = 0;
    var completed = 0;
    final total = ids.length;

    Future<void> worker() async {
      while (index < ids.length) {
        final i = index++;
        await Future<void>.delayed(
          const Duration(milliseconds: _requestDelayMs),
        );
        final value = await fn(ids[i]);
        if (value != null) results.add(value);
        completed++;
        onProgress?.call(completed, total);
      }
    }

    if (ids.isEmpty) {
      onProgress?.call(0, 0);
      return results;
    }

    await Future.wait(
      List.generate(
        ids.length < _concurrency ? ids.length : _concurrency,
        (_) => worker(),
      ),
    );
    return results;
  }

  void dispose() => _client.close();
}

/// キャラ upgrade 同期結果
typedef CharacterUpgradeSync =
    ({
      String characterId,
      List<PromoteStage> promotes,
      Map<String, List<TalentLevelUpgrade>> talents,
    });

/// 武器 upgrade 同期結果
typedef WeaponUpgradeSync =
    ({
      String weaponId,
      List<PromoteStage> promotes,
      List<String> levelUpItemIds,
    });

File: genshin-builder-app/src/lib/api/amber-upgrade.ts
/**
 * Project Amber から突破・天賦・レベルEXPデータを取得する
 */

import { LEVEL_MARKS } from "@/lib/level-config";
import type {
  CharacterUpgradeData,
  LevelExpSegmentData,
  LevelUpMaterialData,
  PromoteStageData,
  TalentLevelUpgradeData,
  TalentUpgradeData,
  WeaponUpgradeData,
} from "./upgrade-types";
import {
  fetchJsonObject,
  UpstreamFetchError,
} from "./safe-json-fetch";

const BASE_URL = "https://gi.yatta.moe";
const REVALIDATE_SEC = 60 * 60 * 24;

const EXP_MATERIAL_IDS = {
  character: ["104001", "104002", "104003"],
  weapon: ["104011", "104012", "104013"],
} as const;

interface ApiTalentPromote {
  level?: number;
  costItems?: Record<string, number> | null;
  coinCost?: number | null;
}

interface ApiTalent {
  type: number;
  name: string;
  promote?: Record<string, ApiTalentPromote>;
}

interface ApiPromote {
  promoteLevel?: number;
  unlockMaxLevel?: number;
  costItems?: Record<string, number>;
  coinCost?: number;
  requiredPlayerLevel?: number;
}

interface ApiAvatarDetail {
  talent: Record<string, ApiTalent>;
  upgrade: { promote?: ApiPromote[] };
}

interface ApiWeaponDetail {
  upgrade: { promote?: ApiPromote[] };
  items?: Record<string, unknown>;
}

interface ApiMaterialDetail {
  name: string;
  description: string;
  type: string;
}

async function fetchJson<T>(path: string): Promise<T> {
  const json = await fetchJsonObject(`${BASE_URL}${path}`, {
    timeoutMs: 20_000,
    maxBytes: 2 * 1024 * 1024,
    retries: 2,
    revalidateSeconds: REVALIDATE_SEC,
  });
  if (
    json.response !== 200 ||
    typeof json.data !== "object" ||
    json.data === null ||
    Array.isArray(json.data)
  ) {
    throw new UpstreamFetchError("invalidData");
  }
  return json.data as T;
}

function parsePromotes(promotes: ApiPromote[] | undefined): PromoteStageData[] {
  return (promotes ?? []).map((promote) => {
    const promoteLevel = promote.promoteLevel ?? 0;
    const unlockMaxLevel = promote.unlockMaxLevel ?? 90;
    const coinCost = promote.coinCost ?? 0;
    const costs = promote.costItems ?? {};
    if (
      promoteLevel < 0 ||
      unlockMaxLevel < 1 ||
      coinCost < 0 ||
      Object.entries(costs).some(
        ([id, count]) =>
          id.trim() === "" ||
          !Number.isInteger(count) ||
          count < 0,
      )
    ) {
      throw new UpstreamFetchError("invalidData");
    }
    return {
      promoteLevel,
      unlockMaxLevel,
      costItems: costs,
      coinCost,
      requiredPlayerLevel: promote.requiredPlayerLevel,
    };
  });
}

function parseTalentUpgrades(
  promote: Record<string, ApiTalentPromote> | undefined,
): TalentLevelUpgradeData[] {
  if (!promote) return [];
  return Object.values(promote)
    .filter((p) => p.level != null)
    .sort((a, b) => (a.level ?? 0) - (b.level ?? 0))
    .map((promote) => {
      const level = promote.level!;
      const coinCost = promote.coinCost ?? 0;
      const costs = promote.costItems ?? {};
      if (
        level < 1 ||
        coinCost < 0 ||
        Object.entries(costs).some(
          ([id, count]) =>
            id.trim() === "" ||
            !Number.isInteger(count) ||
            count < 0,
        )
      ) {
        throw new UpstreamFetchError("invalidData");
      }
      return { level, costItems: costs, coinCost };
    });
}

/** キャラクターの突破・天賦強化データを取得 */
export async function fetchCharacterUpgradeFromApi(
  characterId: string,
): Promise<CharacterUpgradeData> {
  const data = await fetchJson<ApiAvatarDetail>(`/api/v2/jp/avatar/${characterId}`);
  if (
    !data.talent ||
    typeof data.talent !== "object" ||
    !data.upgrade ||
    typeof data.upgrade !== "object" ||
    !Array.isArray(data.upgrade.promote)
  ) {
    throw new UpstreamFetchError("invalidData");
  }

  const entries = Object.keys(data.talent)
    .sort((a, b) => Number(a) - Number(b))
    .map((key) => data.talent[key]);

  const active = entries.filter((t) => t.type === 0 || t.type === 1);
  const activeKinds: Array<"normal" | "skill" | "burst"> = [
    "normal",
    "skill",
    "burst",
  ];

  const talents: TalentUpgradeData[] = active
    .slice(0, 3)
    .map((t, i) => ({
      kind: activeKinds[i],
      upgrades: parseTalentUpgrades(t.promote),
    }))
    .filter((t) => t.upgrades.length > 0);

  const promotes = parsePromotes(data.upgrade.promote);
  if (promotes.length === 0) {
    throw new UpstreamFetchError("invalidData");
  }
  return {
    characterId,
    promotes,
    talents,
  };
}

/** 武器の突破・レベルアップ素材IDを取得 */
export async function fetchWeaponUpgradeFromApi(
  weaponId: string,
): Promise<WeaponUpgradeData> {
  const data = await fetchJson<ApiWeaponDetail>(`/api/v2/jp/weapon/${weaponId}`);
  if (
    !data.upgrade ||
    typeof data.upgrade !== "object" ||
    !Array.isArray(data.upgrade.promote)
  ) {
    throw new UpstreamFetchError("invalidData");
  }

  const levelUpItemIds = EXP_MATERIAL_IDS.weapon.filter(
    (id) => data.items?.[id] != null,
  );

  const promotes = parsePromotes(data.upgrade.promote);
  if (promotes.length === 0) {
    throw new UpstreamFetchError("invalidData");
  }
  return {
    weaponId,
    promotes,
    levelUpItemIds:
      levelUpItemIds.length > 0 ? [...levelUpItemIds] : [...EXP_MATERIAL_IDS.weapon],
  };
}

/** 素材詳細から経験値をパース（description 内の「経験値XXXX」） */
export function parseExpFromDescription(description: string): number | null {
  const match = description.match(/経験値(\d+)/);
  return match ? Number(match[1]) : null;
}

/** 経験値素材の詳細をAPIから取得 */
export async function fetchLevelUpMaterialsFromApi(
  onFetch?: () => void,
): Promise<LevelUpMaterialData[]> {
  const result: LevelUpMaterialData[] = [];

  for (const targetType of ["character", "weapon"] as const) {
    for (const id of EXP_MATERIAL_IDS[targetType]) {
      onFetch?.();
      const detail = await fetchJson<ApiMaterialDetail>(
        `/api/v2/jp/material/${id}`,
      );
      if (
        typeof detail.name !== "string" ||
        detail.name.trim() === "" ||
        typeof detail.description !== "string" ||
        typeof detail.type !== "string"
      ) {
        throw new UpstreamFetchError("invalidData");
      }
      const exp = parseExpFromDescription(detail.description);
      if (exp == null || exp <= 0) {
        throw new UpstreamFetchError("invalidData");
      }
      result.push({
        materialId: id,
        name: detail.name,
        exp,
        targetType,
      });
    }
  }

  if (result.length !== 6) {
    throw new UpstreamFetchError("invalidData");
  }
  return result;
}

/**
 * 目盛り間必要EXP（Amber API に直接のエンドポイントがないため、
 * ゲーム内実測値を同期時にDBへ保存する。値は community verified data）
 */
export function buildLevelExpSegments(): LevelExpSegmentData[] {
  const characterExp: Record<string, number> = {
    "1-20": 12_275,
    "20-30": 57_900,
    "30-40": 65_700,
    "40-50": 39_300,
    "50-60": 94_800,
    "60-70": 114_300,
    "70-80": 280_800,
    "80-90": 393_750,
  };

  const weaponExpByRarity: Record<number, Record<string, number>> = {
    3: {
      "1-20": 53_475,
      "20-30": 127_978,
      "30-40": 145_247,
      "40-50": 275_350,
      "50-60": 408_650,
      "60-70": 572_725,
      "70-80": 772_825,
      "80-90": 1_638_650,
    },
    4: {
      "1-20": 81_000,
      "20-30": 194_512,
      "30-40": 220_613,
      "40-50": 418_725,
      "50-60": 618_400,
      "60-70": 866_675,
      "70-80": 1_168_350,
      "80-90": 2_476_475,
    },
    5: {
      "1-20": 121_550,
      "20-30": 291_591,
      "30-40": 331_209,
      "40-50": 628_150,
      "50-60": 927_675,
      "60-70": 1_299_125,
      "70-80": 1_750_375,
      "80-90": 3_714_775,
    },
  };

  const segments: LevelExpSegmentData[] = [];

  for (let i = 0; i < LEVEL_MARKS.length - 1; i++) {
    const from = LEVEL_MARKS[i];
    const to = LEVEL_MARKS[i + 1];
    const key = `${from}-${to}`;
    const charExp = characterExp[key] ?? 0;
    segments.push({
      id: `character-0-${from}-${to}`,
      targetType: "character",
      rarity: 0,
      fromLevel: from,
      toLevel: to,
      expRequired: charExp,
      moraRequired: Math.round(charExp / 10),
    });

    for (const rarity of [3, 4, 5]) {
      const exp = weaponExpByRarity[rarity][key] ?? 0;
      segments.push({
        id: `weapon-${rarity}-${from}-${to}`,
        targetType: "weapon",
        rarity,
        fromLevel: from,
        toLevel: to,
        expRequired: exp,
        moraRequired: Math.round(exp / 10),
      });
    }
  }

  return segments;
}

/** 並列数を制限して配列を処理（欠損結果は invalidData として拒否） */
export async function mapWithConcurrency<T, R>(
  items: T[],
  concurrency: number,
  fn: (item: T) => Promise<R>,
): Promise<R[]> {
  if (items.length === 0) return [];

  const results: R[] = new Array(items.length);
  let index = 0;

  async function worker() {
    while (true) {
      const i = index++;
      if (i >= items.length) return;
      results[i] = await fn(items[i]);
    }
  }

  await Promise.all(
    Array.from({ length: Math.min(concurrency, items.length) }, worker),
  );
  if (results.some((value) => value === undefined)) {
    throw new UpstreamFetchError("invalidData");
  }
  return results;
}
